src/librustc_mir/transform/validate.rs

```diff
@@ -26,12 +31,16 @@ impl<'tcx> MirPass<'tcx> for Validator {
     fn run_pass(&self, tcx: TyCtxt<'tcx>, source: MirSource<'tcx>, body: &mut Body<'tcx>) {
         let def_id = source.def_id();
         let param_env = tcx.param_env(def_id);
-        TypeChecker { when: &self.when, def_id, body, tcx, param_env }.visit_body(body);
+        let validating_shim =
+            if let ty::InstanceDef::Item(_) = source.instance { false } else { true };
```

Comment:
Oh, that's interesting, I thought this would be passed in. But I guess this works, too? I am not very familiar with `InstanceDef`.

Reply:
Have we found any instances where something that is *not* `Self` is being called? Because then you can also special-case this even further by checking that the callee is `ty.is_param(0)`, instead of doing a trait query. And we could also *only* special-case the specific call shims we care about.

Reply:
I don't think we have, but I think that this approach is a little more liberal in the sense that it allows shim implementations to change without changing the assumptions we require here - there's definitely a trade-off here though

Reply:
Shims shouldn't grow more ways to violate the assumptions we make here, they should only change so they use an `FnDef` instead of `Self` as the callee eventually, fixing the underlying issue

Reply:
I mean to say that another class of shim could be introduced or a previously monomorphic shim could be made polymorphic

Reply:
I agree it's tidier `ty.is_param(0)`, but we really would like `Self` to implement `FnOnce` right? It would almost certainly be malformed MIR if it didn't

Reply:
All I can think at the moment is that we include both checks?

Reply:
That would work too

Reply:
Ah, you can just compare with `tcx.types.self_param` too